File: plugins/_shared/scripts/edit_targets.py
#!/usr/bin/env python3
"""Edit target extraction for Claude/Codex hook payloads."""

import os
import re
from typing import Optional


def iter_edit_targets(data: dict) -> list[dict]:
    """Return Write/Edit/apply_patch targets from hook input."""
    tool = data.get("tool_name")
    tool_input = data.get("tool_input") or {}
    if tool == "apply_patch":
        command = str(tool_input.get("command")
                      or tool_input.get("input")
                      or tool_input.get("patch") or "")
        return _parse_apply_patch(command)
    if tool not in ("Write", "Edit"):
        return []
    file_path = str(tool_input.get("file_path") or "")
    if not file_path:
        return []
    content = str(tool_input.get("content") or tool_input.get("new_string") or "")
    return [{"file_path": file_path, "content": content, "action": tool}]


def first_edit_target(data: dict, file_pattern: str,
                      skip_pattern: str = "") -> Optional[dict]:
    """Return first edited target matching file and skip filters."""
    for target in iter_edit_targets(data):
        file_path = target.get("file_path", "")
        if not re.search(file_pattern, file_path):
            continue
        if skip_pattern and re.search(skip_pattern, file_path):
            continue
        return target
    return None


def _parse_apply_patch(command: str) -> list[dict]:
    """Parse Codex apply_patch V4A command into edited file targets."""
    targets = []
    file_path = ""
    action = ""
    added = []

    def flush() -> None:
        if file_path and action != "Delete":
            targets.append({
                "file_path": file_path,
                "content": "\n".join(added),
                "action": action,
            })

    for line in command.splitlines():
        match = re.match(r"^\*\*\*\s+(Add|Update|Delete) File:\s+(.+)$", line)
        if match:
            flush()
            action = match.group(1)
            file_path = match.group(2).strip()
            added = []
            continue
        if action in ("Add", "Update") and line.startswith("+") and not line.startswith("+++"):
            added.append(line[1:])
    flush()
    return targets


def read_existing_content(file_path: str) -> str:
    """Read a file if it exists, returning an empty string on failure."""
    if not os.path.isfile(file_path):
        return ""
    try:
        with open(file_path, encoding="utf-8") as f:
            return f.read()
    except OSError:
        return ""


def target_content(target: dict, prefer_existing: bool = False) -> str:
    """Return useful content for validation from an edit target."""
    file_path = str(target.get("file_path") or "")
    content = str(target.get("content") or "")
    if prefer_existing:
        return read_existing_content(file_path) or content
    return content

File: plugins/ai-pilot/scripts/lib/apex/edit-targets.ts
import type { HookInput } from "../interfaces/hook.interface";
import { asRecord } from "../../../../core-guards/scripts/_shared/as-record";

export type EditTarget = {
  filePath: string;
  content: string;
  isTrivialEdit: boolean;
};

function parseApplyPatch(body: string): EditTarget[] {
  const targets: EditTarget[] = [];
  let filePath = "";
  let action = "";
  let added: string[] = [];

  const flush = (): void => {
    if (!filePath || action === "Delete") return;
    targets.push({ filePath, content: added.join("\n"), isTrivialEdit: action === "Update" && added.length < 5 });
  };

  for (const line of body.split("\n")) {
    const match = line.match(/^\*\*\*\s+(Add|Update|Delete) File:\s+(.+)$/);
    if (match) {
      flush();
      action = match[1] ?? "";
      filePath = (match[2] ?? "").trim();
      added = [];
    } else if ((action === "Add" || action === "Update") && line.startsWith("+") && !line.startsWith("+++")) {
      added.push(line.slice(1));
    }
  }
  flush();
  return targets;
}

export function editTargets(input: HookInput | unknown): EditTarget[] {
  const data = asRecord(input);
  if (!data) return [];
  const toolName = typeof data.tool_name === "string" ? data.tool_name : "";
  const toolInput = asRecord(data.tool_input) ?? {};
  if (toolName === "apply_patch") {
    return parseApplyPatch(String(toolInput.command ?? toolInput.input ?? toolInput.patch ?? ""));
  }
  if (toolName !== "Write" && toolName !== "Edit") return [];
  const filePath = String(toolInput.file_path ?? "");
  if (!filePath) return [];
  const content = String(toolInput.content ?? toolInput.new_string ?? "");
  const replaceAll = Boolean(toolInput.replace_all);
  return [{
    filePath,
    content,
    isTrivialEdit: toolName === "Edit" && !replaceAll && content.split("\n").length < 5,
  }];
}
